odoo/models.py

```diff
@@ -5513,64 +5513,73 @@ def filtered(self, func):
+                    result.append(self.search([('id', 'in', self.ids), leaf], order='id'))
                     continue
+
                 if key.endswith('.id'):
                     key = key[:-3]
                 if key == 'id':
                     key = ''
+
                 # determine the field with the final type for values
                 field = None
                 if key:
                     model = self.browse()
                     for fname in key.split('.'):
                         field = model._fields[fname]
                         model = model[fname]
+
                 if comparator in ('like', 'ilike', '=like', '=ilike', 'not ilike', 'not like'):
                     value_esc = value.replace('_', '?').replace('%', '*').replace('[', '?')
+                if comparator in ('in', 'not in'):
+                    if isinstance(value, (list, tuple)):
+                        value = set(value)
+                    else:
+                        value = (value,)
+                    if field and field.type in ('date', 'datetime'):
+                        value = {Datetime.to_datetime(v) for v in value}
+                elif field and field.type in ('date', 'datetime'):
+                    value = Datetime.to_datetime(value)
+
                 records_ids = OrderedSet()
                 for rec in self:
                     data = rec.mapped(key)
                     if isinstance(data, BaseModel):
                         v = value
-                        if (isinstance(value, list) or isinstance(value, tuple)) and len(value):
-                            v = value[0]
+                        if isinstance(value, (list, tuple, set)) and value:
+                            v = next(iter(value))
                         if isinstance(v, str):
                             data = data.mapped('display_name')
                         else:
                             data = data and data.ids or [False]
                     elif field and field.type in ('date', 'datetime'):
-                        # convert all date and datetime values to datetime
-                        normalize = Datetime.to_datetime
-                        if isinstance(value, (list, tuple)):
-                            value = [normalize(v) for v in value]
-                        else:
-                            value = normalize(value)
-                        data = [normalize(d) for d in data]
-                    if comparator in ('in', 'not in'):
-                        if not (isinstance(value, list) or isinstance(value, tuple)):
-                            value = [value]
+                        data = [Datetime.to_datetime(d) for d in data]
 
                     if comparator == '=':
                         ok = value in data
                     elif comparator == 'in':
-                        ok = any(map(lambda x: x in data, value))
+                        ok = any(map(lambda x: x in value, data))
```

Comment:
This `any(map(...))` is non-pythonic, simply use a comprehension like `any(x in value for x in data)`